test/unified_format.py

```diff
@@ -0,0 +1,283 @@
+            # TODO
+            # Add logic to respect the following fields
+            # - collectionOptions
+            database = self.entity_map[spec['database']]
+            self.entity_map[spec['id']] = database.get_collection(spec['collectionName'])
+            return
+        # elif ...
+            # TODO
+            # Implement the following entity types:
+            # - session
+            # - bucket
+        else:
+            raise ValueError("Unknown entity type %s" % (entity_type,))
+
+    def setUp(self):
+        super(UnifiedSpecTestMixin, self).setUp()
+
+        # process createEntities
+        self.entity_map = {}
+        entity_spec = self.TEST_SPEC.get('createEntities', [])
+        for spec in entity_spec:
+            self.create_entity(spec)
+
+        # process initialData
+        initial_data = self.TEST_SPEC.get('initialData', [])
+        for data in initial_data:
+            self.insert_initial_data(self.client, data)
+
+        # initialize internals
+        self.test_assets = {}
+
+    def run_scenario(self, spec):
+        # process test-level runOnRequirements
+        run_on_spec = spec.get('runOnRequirements', [])
+        if not self.should_run_on(run_on_spec):
+            raise unittest.SkipTest('runOnRequirements not satisfied')
+
+        # process skipReason
+        skip_reason = spec.get('skipReason', None)
+        if skip_reason is not None:
+            raise unittest.SkipTest('%s' % (skip_reason,))
+
+        # process operations
+        # TODO: process operations
+
+        # process expectedEvents
+        # TODO: process expectedEvents
+
+        # process outcome
+        # TODO: process outcome
+
+
+class UnifiedSpecTestMeta(type):
+    """Metaclass for generating test classes."""
+    def __init__(cls, *args, **kwargs):
+        super(UnifiedSpecTestMeta, cls).__init__(*args, **kwargs)
+
+        def create_test(spec):
+            def test_case(self):
+                self.run_scenario(spec)
+            return test_case
+
+        for test_spec in cls.TEST_SPEC['tests']:
+            description = test_spec['description']
+            test_name = 'test_%s' % (
+                description.replace(' ', '_').replace('.', '_'),)
+            test_method = create_test(copy.deepcopy(test_spec))
+            test_method.__name__ = test_name
+            setattr(cls, test_name, test_method)
+
+
+def generate_test_classes(test_path, module=__name__, class_name_prefix=''):
+    """Method for generating test classes. Returns a dictionary where keys are
+    the names of test classes and values are the test class objects."""
+    test_klasses = {}
+
+    def test_base_class_factory(test_spec):
+        """Utility that creates the base class to use for test generation.
+        This is needed to ensure that cls.TEST_SPEC is appropriately set when
+        the metaclass __init__ is invoked."""
+        class SpecTestBase(with_metaclass(UnifiedSpecTestMeta)):
```

Comment:
I'm interested to know why you chose to use a meta class here. It seems like we could get the same result by just creating a new class and adding the tests like we currently do:
```
class TestUnified(UnifiedSpecTestMixin):pass
for test_spec in scenario_def['tests']:
     ...
     setattr(TestUnified, test_name, test_method)
```
Am I missing something?

Reply:
I don't think you are missing anything - we could certainly go that route and get the same result. I simply started with the metaprogramming approach because in my mind, defining classes programmatically is precisely what metaclasses are for. 
Metaclasses do offer a high degree of customizability and control over class creation and though we aren't currently doing anything that cannot be achieved without metaclasses, this might change in the future depending on how the unified format changes and where we end up using the runner for it. In this light, I think this metaclass based architecture will be more flexible/future-proof.

Reply:
Take for instance the scenario when the next major `schemaVersion` of the unified format is introduced. Using metaclasses we can easily customize which Mixin class to use based on the `schemaVersion` of the file. Again, we can also do this without metaclasses though it would be less elegant.

Reply:
SGTM.

Reply:
🥳